### PR TITLE
TFMC updates + unbinned v6 fits w/ backgrounds

### PR DESCRIPTION
**Pull request by rbarrue**
In view of upcoming v6 fits with backgrounds, I updated the TFMC code to include the train and validation split used for the PNNs. I mostly took the code from pnn_training.py, updated to the TFMC specifics.

The original TFMC code builds the training sample by combining the events from each class into a single sample (keeping class labels) then shuffling them. To do something similar taking into account the train-validation split, I take the training fraction of events from each sample and then combine them into a single training sample, which I then shuffle. I do the same thing for the validation. My goal was to keep the same relative fraction of events from each class in both samples.

~To get the inclusive cross-sections, which we need to convert classifier output probabilities into differential cross-section ratios, the original TFMC code iterated through the entire sample and accumulated the weights for each class. Now that we are no longer using the entire sample, I modified the code to use the training fraction of events to accumulate the weights. It will not give us the correct values, but it should keep the relative cross-section differences correct, which in my mind should lead to the correct DCR. I will double-check this with pen and paper and if I find something I'll change it. One way to get closer to the correct values would be to divide the obtained values by the training partition fraction. This relies on the splitting and shuffling being random enough that a fraction _x_ of the sample has a sum of weights = ~ _x_*full cross-section.~ Changed the code to get the cross-sections accumulating from the entire sample, but only use the training fraction for training.

WIP so keeping as draft for now.

### CODE REVIEW

**Comment by rbarrue** [comment]
Main items to still be implemented:
- implement access to last epoch and best epochs (via text files)
- implement early stopping (will require changing yaml_loader._apply_defaults_and checks such that the default can be used)

File: ML/TFMC/tfmc_training.py [hunk 1]
```diff
@@ -129,6 +128,45 @@ def list_jobs_and_exit():
...
+        n_buckets=uid_n_buckets,
+    )
+
+    # build bucket intervals (inline; no extra helper, no extra checks)
+    keys  = list(uid_scheme.keys())
+    fracs = [float((uid_scheme[k] or {}).get("fraction", 0.0)) for k in keys]
+
+    sizes = [int(math.floor(f * uid_n_buckets)) for f in fracs]
+    sizes[-1] += uid_n_buckets - sum(sizes)
+
```
[the remark this answers is not in the document]
Comment (rbarrue) [reply ML/TFMC/tfmc_training.py:175]: Eh, will ignore this for now, will fix it if it later if it happens to be necessary.

File: ML/TFMC/tfmc_training.py [hunk 2]
```diff
@@ -129,6 +128,45 @@ def list_jobs_and_exit():
...
+    uid_intervals = {}
+    lo = 0
+    for k, sz in zip(keys, sizes):
+        uid_intervals[k] = (lo, lo + int(sz))
+        lo += int(sz)
+    pnn_train_key = "pnn_train"
+    pnn_val_key = "pnn_val"
+    train_interval = uid_intervals[pnn_train_key]
+    val_interval   = uid_intervals[pnn_val_key]
```
[the remark this answers is not in the document]
Comment (rbarrue) [reply ML/TFMC/tfmc_training.py]: Also won't touch this for now.

File: ML/TFMC/tfmc_training.py [hunk 3]
```diff
@@ -138,23 +176,94 @@ def iterate_epoch(shard_limit: int | None = None):
...
-        w = np.concatenate(Ws, axis=0) if Ws else np.empty((0,))
-        #yield X, y, w
-        #idx = rng.permutation(len(X))
-        #yield X[idx], y[idx], w[idx]
-        idx = np.random.permutation(len(X))
-        yield X[idx], y[idx], w[idx]
+
+            if uid_enabled:
+                # training and validation events from UID-based splitting mask
+                obs_names = L.observer_names
```
[the remark this answers is not in the document]
Comment (rbarrue) [reply ML/TFMC/tfmc_training.py:217]: Won't touch this for now, since our samples are all produced in the same way and have those fields.

**Comment by rbarrue** [comment]
Implemented a flag to do train-val splitting in iterate_epoch to allow cross-section calculation (+ yield check) done with all the events, as previously, but keeping the train-val split in the training.

File: ML/TFMC/tfmc_training.py [hunk 4]
```diff
@@ -355,43 +472,73 @@ def plot_convergence_root(true_h, pred_h, epoch, out_dir, feature_names, classes
...
+                start_val, stop_val = math.ceil(start*val_tr_size_ratio), min(math.ceil(stop*val_tr_size_ratio),N_val)
+                Xb_val, yb_val, wb_val = X_val[start_val:stop_val], y_val[start_val:stop_val], w_val[start_val:stop_val]
+                
+                # train_on_batch performs weight updates
+                loss_train = model.train_on_batch(Xb_tr, yb_tr, wb_tr)
+                losses_train.append(loss_train)
+
+                # compute_loss computes loss without gradient updates
+                loss_val = model.compute_loss(Xb_val, yb_val, wb_val)
+                losses_val.append(loss_val)
```
[the remark this answers is not in the document]
Comment (rbarrue) [reply ML/TFMC/tfmc_training.py]: This is a corner cases that will only happen for unrealistically small batch sizes, so will not implement this here.

**Comment by rbarrue** [comment]
Given the issues seen with convergence in the presence of a large class imbalance (both in terms of weighted and unweighted numbers of events) implementing a new feature: give the inclusive cross-section ratios as prior probabilities (via softmax logits) and have the network learn around that.

First core implementation is in with hardcoded event number ratios, the next step is to connect it with the ratio calculation done to get the class weights.

**Comment by rbarrue** [comment]
The implementation of setting the logit priors to the inclusive XS ratio is postponed, as we are now experimenting with simpler ways to have stable trainings. To avoid noise in the code, I removed that in commit https://github.com/HephyAnalysisSW/GOLLUM/pull/122/commits/2060baa2eaa4e4020af7a2a44d4a137ff2f300d0

We can return to that if we have exhausted all other options.

**Comment by rbarrue** [comment]
The main new feature implemented into TFMC training is Early Stopping. @Dorhand can you have a look ?

**Comment by rbarrue** [comment]
Additionally, I'm also storing the last and best epoch information. This will allow us to start trainings from the last epoch for cases where we e.g. ran into OOM or max. walltime. 

I also implemented a feature both in the TFMC and PNN training to avoid restarting trainings which have properly finished, without using the `--overwrite` flag.

PS: I also implemented access to RDataLoader n_split access in TFMC training.

File: ML/TFMC/tfmc_training.py [hunk 5]
```diff
@@ -129,29 +134,120 @@ def list_jobs_and_exit():
...
         n_shards = min(n_shards, shard_limit)
     for shard in range(n_shards):
         Xs, Ys, Ws = [], [], []
+        Xs_tr, Ys_tr, Ws_tr = [], [], []
+        Xs_val, Ys_val, Ws_val = [], [], []
         for ci, (name, L) in enumerate(zip(classes_names, loaders)):
             # Use materialize to fetch features and weights; views apply masks internally.
-            X, w = L.materialize(shard=shard, what="fw", n=None)
+            X, O, w = L.materialize(shard=shard, what="fow", n=None)
             y = np.zeros((len(X), len(classes_names)), dtype=np.float32)
```
[the remark this answers is not in the document]
Comment (rbarrue) [reply ML/TFMC/tfmc_training.py:212]: Will not change this, the I/O cost is not that serious.

File: ML/TFMC/tfmc_training.py [hunk 6]
```diff
@@ -341,57 +441,155 @@ def plot_convergence_root(true_h, pred_h, epoch, out_dir, feature_names, classes
...
+def _is_improved(current: float, best: float) -> bool:
+    # min: want current < best - min_delta
+    # max: want current > best + min_delta
+    if mode == "min":
+        return current < (best - min_delta)
+    else:
+        return current > (best + min_delta)
+best_val = float("inf")
+best_epoch = -1
+bad_epochs = 0
```
[the remark this answers is not in the document]
Comment (rbarrue) [reply ML/TFMC/tfmc_training.py]: I fixed your comment by copying what was in the PNN training script.

File: ML/TFMC/TFMC.py [hunk 7]
```diff
@@ -66,16 +66,15 @@ def __init__(
...
         m = Sequential()
         m.add(layers.Input(shape=(self.input_dim,)))
         for units in self.hidden_layers:
-            m.add(layers.Dense(units, activation=None, kernel_regularizer=reg))
-            m.add(layers.Activation(self.activation))
+            m.add(layers.Dense(units, activation=self.activation, kernel_regularizer=reg))
             if self.dropout_rate and self.dropout_rate > 0:
                 m.add(layers.Dropout(self.dropout_rate))
-        m.add(layers.Dense(self.num_classes, activation="softmax", kernel_regularizer=reg))
+            m.add(layers.Dense(self.num_classes, activation="softmax", kernel_regularizer=reg))
```
[the remark this answers is not in the document]
Comment (rbarrue) [reply ML/TFMC/TFMC.py]: Good catch. Fixed.
